File: src/java/first_bad_version.java
/* The isBadVersion API is defined in the parent class VersionControl.
boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int i = 1;
        int j = n;
        
        int ans = j;
        while (i <= j) {
            int m = i + (j-i)/2;
            if (isBadVersion(m)) {
                ans = m;
                j=m-1;
            }
            else {
                i=m+1;
            }
        }
        return ans;
    }
}
// Submission Link: https://leetcode.com/problems/first-bad-version/submissions/634890233/

File: src/cpp/first-bad-version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int start = 1;
        int end = n;
        int mid = start+(end-start)/2;
        while(start<end){
            if(isBadVersion(mid)){
                end = mid;
            }
            else {
                start = mid+1;
            }
            mid = start+(end-start)/2;
        }
        return mid;
    }
};